### PR TITLE
Add BundleAccessorOptions

### PR DESCRIPTION
**Pull request by danibachar**
Resolves https://github.com/tuist/tuist/issues/6149

### Short description 📝

Adding BundleAccessorOptions. 
Currently bundle accessors generation is solely defined using a boolean flag and both acessors for objective c and swift runtime are generated. 

The fact is that the objective c accessors is not really needed al the time, moreover when the class symbols can class with symbols that already existing within a target.

The suggestion here is an additional layer which allows a more fine-grain control over the code generation.

I tried and keeping backward compatibility as much as possible, and as such introduce some new initialization function

### How to test the changes locally 🧐

> Include a set of steps for the reviewer to test the changes locally (see [the documentation](https://docs.tuist.io/contributors/testing-strategy) for reference).

### Contributor checklist ✅

- [x] The code has been linted using run `mise run lint:fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/ProjectDescription/ProjectOptions.swift [hunk 1]
```diff
@@ -36,12 +36,38 @@ extension Project {
...
+                developmentRegion: developmentRegion,
+                bundleAccessorsOptions: bundleAccessorsOptions,
+                disableShowEnvironmentVarsInScriptPhases: disableShowEnvironmentVarsInScriptPhases,
+                disableSynthesizedResourceAccessors: disableSynthesizedResourceAccessors,
+                textSettings: textSettings,
+                xcodeProjectName: xcodeProjectName
+            )
+        }
+
+        public static func projectOptions(
```
Comment (danibachar) [inline Sources/ProjectDescription/ProjectOptions.swift:56]: I'm not sure if we want to deprecate the old initializers, I had to add an additional one to keep backward compatibility and allow compilation (default values in the functions caused a compilation issue)

File: Sources/TuistGraph/Models/ProjectOptions.swift [hunk 2]
```diff
@@ -26,7 +26,25 @@ extension Project {
...
+                bundleAccessorsOptions = .disabled
+            } else {
+                bundleAccessorsOptions = .enabled(includeObjcAccessor: true)
+            }
+            self.disableShowEnvironmentVarsInScriptPhases = disableShowEnvironmentVarsInScriptPhases
+            self.disableSynthesizedResourceAccessors = disableSynthesizedResourceAccessors
+            self.textSettings = textSettings
+        }
+
+        public init(
```
Comment (danibachar) [inline Sources/TuistGraph/Models/ProjectOptions.swift:39]: I'm not sure if we want to deprecate old initializers, I had to add an additional one to keep backward compatibility and allow compilation (default values in the functions caused a compilation issue)

File: Sources/TuistGraphTesting/Models/ProjectOptions+TestData.swift [hunk 3]
```diff
@@ -25,6 +29,22 @@ extension Project.Options {
             textSettings: textSettings
         )
     }
+
+    public static func testOptions(
```
Comment (danibachar) [inline Sources/TuistGraphTesting/Models/ProjectOptions+TestData.swift:33]: I'm not sure if we want to deprecate old initializers, I had to add an additional one to keep backward compatibility and allow compilation (default values in the functions caused a compilation issue)

I can definitely add a deprecation warning if we decide to adopt this API

File: Sources/TuistLoader/Models+ManifestMappers/ProjectOption+ToManifestMapper.swift [hunk 4]
```diff
@@ -6,9 +6,9 @@ extension ProjectDescription.Project.Options {
     /// - Parameters:
     /// - manifest: Manifest representation of project options.
     static func from(manifest: TuistGraph.Project.Options) -> Self {
-        options(
+        projectOptions(
```
Comment (danibachar) [inline Sources/TuistLoader/Models+ManifestMappers/ProjectOption+ToManifestMapper.swift:9]: I'm not sure if we want to deprecate old initializers, I had to add an additional one to keep backward compatibility and allow compilation (default values in the functions caused a compilation issue)

I can definitely add a deprecation warning if we decide to adopt this API

File: Sources/ProjectDescription/ProjectOptions.swift [hunk 5]
```diff
@@ -12,8 +12,8 @@ extension Project {
         /// Configures the development region.
         public var developmentRegion: String?
 
-        /// Disables generating Bundle accessors.
-        public var disableBundleAccessors: Bool
+        /// Defines if and how bundle accessors are generated
+        public var bundleAccessorsOptions: BundleAccessorOptions
```
Comment (danibachar) [inline Sources/ProjectDescription/ProjectOptions.swift:16]: This is a public API breaking change - while I think its ok, I can keep support for the boolean flag.
I was mainly thinking it is checked internally and set externally and as such didn't see an issue removing it

**Comment by danibachar** [comment]
I didn't tick the documentation part, as I couldn't find any specific ones around this API. I'm guessing as this is an API change the API docs are getting updated automatically and this will be addressed there

**Comment by danibachar** [comment]
Some images of before/after

Before - Framework


Screenshot 2024-03-26 at 11 19 58 AM

After - Framework

Screenshot 2024-03-26 at 11 19 58 AM

**Review by fortmarek — COMMENTED** [review]
Hey 👋 

First of all, thanks a lot for the contribution. However, I'm not sure if more fine-grained control is actually fixing the core issue – which is generating a new class with a pre-baked name that's public.

From what I can see, a better solution might be:
- change the objc resource accessor to also do an extension on `Foundation.Bundle` as we do for the Swift accessor. I'm not 100 % sure why the objc resource accessor is that different from the Swift one. I believe one should still be able to call the `module` accessor from objc even if it's in an extension
- make the objc resource accessor not public since the Swift one is also internal

I honestly don't know much about why the current objc resource accessor looks the way it does, maybe @pepicrft might have more context as he added this a long time ago: https://github.com/tuist/tuist/pull/1567

File: Sources/ProjectDescription/ProjectOptions.swift (hunk 5 above)
Comment (pepicrft) [reply Sources/ProjectDescription/ProjectOptions.swift:16]: Let's keep the old one, mark it as deprecated to tell users to use `bundleAccessorOptions`, and default to `bundleAccessorOptions` with a fallback on `disableBundleAccessors`. We've recently gone through a handful of breaking changes with Tuist 4 and we should give users some breaking changes calm.

**Comment by pepicrft** [comment]
> Hey 👋
> 
> First of all, thanks a lot for the contribution. However, I'm not sure if more fine-grained control is actually fixing the core issue – which is generating a new class with a pre-baked name that's public.
> 
> From what I can see, a better solution might be:
> 
>     * change the objc resource accessor to also do an extension on `Foundation.Bundle` as we do for the Swift accessor. I'm not 100 % sure why the objc resource accessor is that different from the Swift one. I believe one should still be able to call the `module` accessor from objc even if it's in an extension
> 
>     * make the objc resource accessor not public since the Swift one is also internal
> 
> 
> I honestly don't know much about why the current objc resource accessor looks the way it does, maybe @pepicrft might have more context as he added this a long time ago: #1567

I just read this comment. I believe it's impossible to create an extension on an Objective-C class, hence creating a new class. Since we are already using internal for the Swift side, I'm inclined to make the Objective-C part internal, too, and avoid the introduction of new APIs. Since there aren't many people with Objective-C (my guess), I think the change shouldn't be that breaking, so we can afford the introduction.

**Comment by fortmarek** [comment]
> I believe it's impossible to create an extension on an Objective-C class, hence creating a new class. Since we are already using internal for the Swift side, I'm inclined to make the Objective-C part internal, too, and avoid the introduction of new APIs

I don't think this is true. I just tested this in the `ios_app_with_framework_and_resources` fixture and when I rewrite the generated objc resource to:
```swift
@objc public class var frameworkResourcesBundle: Bundle {
  .module
}

// Instead of:
// @objc static let module: Bundle = {
```

I can then use the `module` in an objc `.m` file:
```
#import <StaticFramework3/StaticFramework3-Swift.h>

@implementation ObjectiveFeature
- (NSString *)myCustomMethod {
    NSBundle* bundle = [NSBundle frameworkResourcesBundle];
    return @"Result from my custom method";
}
@end
```

Unfortunately, the Swift accessor in this case needs to be _public_, so that would mean that it could clash with other accessors which is why we still need the extra `class var` instead of one `module` property`

File: Sources/TuistLoader/Models+ManifestMappers/ProjectOption+ManifestMapper.swift [hunk 6]
```diff
@@ -8,7 +8,7 @@ extension TuistGraph.Project.Options {
     static func from(manifest: ProjectDescription.Project.Options) -> Self {
         .init(
             automaticSchemesOptions: .from(manifest: manifest.automaticSchemesOptions),
-            disableBundleAccessors: manifest.disableBundleAccessors,
+            bundleAccessorsOptions: manifest.bundleAccessorsOptions.asTuistGraphComponent(),
```
[the remark this answers is not in the document]
Comment (danibachar) [reply Sources/TuistLoader/Models+ManifestMappers/ProjectOption+ManifestMapper.swift:11]: Fixed

File: Sources/ProjectDescription/ProjectOptions.swift [hunk 7]
```diff
@@ -112,3 +138,17 @@ extension Project.Options {
...
+    /// Defines if and how bundle accessors are generated
+    public enum BundleAccessorOptions: Codable, Hashable {
+        /// Enables generated bundle accessors
+        /// Option to control wether an accessor for Objective-C run time will be added as well
+        case enabled(includeObjcAccessor: Bool)
+
+        /// Disables generated bundle accessors
+        case disabled
+    }
+}
```
[the remark this answers is not in the document]
Comment (danibachar) [reply Sources/ProjectDescription/ProjectOptions.swift:154]: Fixed, love the suggestion!

File: Sources/TuistGenerator/Mappers/ResourcesProjectMapper.swift [hunk 8]
```diff
@@ -165,9 +192,15 @@ public class ResourcesProjectMapper: ProjectMapping { // swiftlint:disable:this
     }
 
     // swiftlint:disable:next function_body_length
-    static func fileContent(targetName: String, bundleName: String, target: Target) -> String {
+    static func fileContent(
+        targetName: String,
+
```
[the remark this answers is not in the document]
Comment (danibachar) [reply Sources/TuistGenerator/Mappers/ResourcesProjectMapper.swift:197]: fixed

File: Tests/TuistGeneratorTests/ProjectMappers/ResourcesProjectMapperTests.swift [hunk 9]
```diff
@@ -382,4 +382,86 @@ final class ResourcesProjectMapperTests: TuistUnitTestCase {
...
+            product: .staticLibrary,
+            sources: ["/Absolute/File.swift"],
+            resources: resources
+        )
+        project = Project.test(
+            options: .testOptions(bundleAccessorsOptions: .enabled(includeObjcAccessor: false)),
+            targets: [target]
+        )
+
+        // Got
```
[the remark this answers is not in the document]
Comment (danibachar) [reply Tests/TuistGeneratorTests/ProjectMappers/ResourcesProjectMapperTests.swift:421]: Fixed, it is a pattern from other tests - I like the BDD approach. Shall I convert the rest in this file?

File: Tests/TuistGeneratorTests/ProjectMappers/ResourcesProjectMapperTests.swift [hunk 10]
```diff
@@ -382,4 +382,86 @@ final class ResourcesProjectMapperTests: TuistUnitTestCase {
...
+            product: .staticLibrary,
+            sources: [.init(stringLiteral: "/Absolute/\(targetName)Resources.swift")],
+            resources: resources
+        )
+        project = Project.test(
+            options: .testOptions(bundleAccessorsOptions: .enabled(includeObjcAccessor: true)),
+            targets: [target]
+        )
+
+        // Got
```
[the remark this answers is not in the document]
Comment (danibachar) [reply Tests/TuistGeneratorTests/ProjectMappers/ResourcesProjectMapperTests.swift:461]: fixed

File: Sources/TuistLoaderTesting/Loaders/TestData/ProjectDescription+TestData.swift [hunk 11]
```diff
@@ -46,7 +46,8 @@ extension Project {
         organizationName: String? = nil,
         settings: Settings? = nil,
         targets: [Target] = [],
-        additionalFiles: [FileElement] = []
+        additionalFiles: [FileElement] = [],
+        options _: Options = .options()
```
[the remark this answers is not in the document]
Comment (danibachar) [reply Sources/TuistLoaderTesting/Loaders/TestData/ProjectDescription+TestData.swift:50]: fixed

**Comment by danibachar** [comment]
I fixed the comments.
I think it is still beneficial to give fin-grain control over code generation - why generating code that is not needed (and of course avoiding symbols clash)?

Having said that - I can work on a second PR that will allow the extension @fortmarek suggested - I like it... My only concern is that it will introduce a breaking change in how users consume that bundle already (i.e old code that uses `[MyTargetResources bundle]`)

Any thoughts/suggestions would be great!

**Comment by pepicrft** [comment]
> I don't think this is true. I just tested this in the ios_app_with_framework_and_resources fixture and when I rewrite the generated objc resource to:

I misunderstood it, then. 

I'm inclined to take a different approach where:
1. We respect the current interface
2. But allow configuring the **name** of the accessor with `module` as a default.
3. Mark the generated Objective-C accessor as deprecated and follow @fortmarek's suggestion for consistency.

What do you think?

**Comment by danibachar** [comment]
In regards to this [comment](https://github.com/tuist/tuist/pull/6124#issuecomment-2027380685), just wanted to make sure I understand your idea.

1) We keep the current API - which allows enable/disable the whole code generation, and not a fine-grain control per accessor.
2) We would still generate the old class name for objc accessor - just with a deprecation warning

Now, in addition for keeping things more or less the same, we want to introduce two changes:

1) change the `static let module: Bundle = {}()` property on the `Foundation.Bundle` extension to be a computed property `@objc public class var module: Bundle {}`
2) We want to give users the ability to change this property name

If I miss-understand - you can stop here, and please explain me what have I missed?

If I understand correctly, I would have several points:

I would start by saying that its probably a personal preference, or a bit of a bias, but for me this feature is most useful when I handle external dependencies, where I don't have control over the source code. This is why fine-grain control is important. For internal modules I can, in worst case, define my own accessors and disable code generation.

1) The fact that we would still generate the old accessor, still pose the issue of clash in symbols (this fine grain control will also support easier depreciation I believe)

2) I'm pretty sure customers will have symbols clash right away with this proposed solution, as we will basically expose `Bundle.module` publicly by default from different modules etc (all of them will expose that same public symbol)

I like the approach of exposing a property on the bundle and giving the user control over that name (as its public).
Maybe we can combine several options? It might end-up in a more cumbersome API?

struct BundleAccessorOptions {
  struct Options: OptionSet {...} // similar to what we have today, to allow control over what API is generated
  let options: Options // default to [.swift, .objc]
  let name: String // will allow control over the property (and possibly the deprecated class name?), and will default to 
}

- We will continue generate the objc class with a deprecated notice - if [.objc] option is set
- We will potentially (depends on the options) generate two accessors like in @fortmarek suggestion swift `static let module: Bundle = {}()`, and an objc `@objc public class var frameworkNameResourcesBundle: Bundle { .module }`. 
- The name will control on both the deprecated class name and property name (no strong feelings about this, can be on or the other, or none). Defaulting the same as today - using the module/framework name as prefix?

WDTY?

**Comment by pepicrft** [comment]
Thanks for your response @danibachar.

> I would start by saying that its probably a personal preference, or a bit of a bias, but for me this feature is most useful when I handle external dependencies, where I don't have control over the source code. This is why fine-grain control is important. For internal modules I can, in worst case, define my own accessors and disable code generation.

Would you mind sharing more about the problem/need that motivated you to make this change in the first place? I feel I'm missing something and therefore my solutions might feel a bit misleading. I read through the body of the issue again, but I couldn't fully extract your motivations from it.

Thanks.

**Comment by danibachar** [comment]
Thank @pepicrft -  the motivation is twofold:
First - A potential symbols clash - the tool is generating a public class - which potentially can clash with other class/struct names in/outside the target/library definition.
Second, efficiency. Why generating code which is not in use? (some static analysis tools will fail CI in such cases).

There are of course other solutions for my two points, but I feel like the tool API could benefit from such fine-grain control over code generation.

**Comment by pepicrft** [comment]
Hi @danibachar 
Thanks for sharing the motivations 🙏🏼. 
While it's tempting to batch present challenges and potentially future challenges and resort to all sorts of configurability that will, in the end, make the interface more complex for users, I'm inclined to focus on the first one and:
- Ensure we are aligned with SPM's work. I believe I used that as a reference when I implemented it back then. I'd check if anything has changed there. 
- Provide an interface to customize the naming, which users can use then to mitigate clashes.

**Comment by fortmarek** [comment]
> Ensure we are aligned with SPM's work. I believe I used that as a reference when I implemented it back then. I'd check if anything has changed there.

+1 on this. Since your use-case is related to an external dependency, we probably shouldn't even have to make the name configurable and rather follow Apple's lead on this one.

The best way to check what Apple is doing is to reproduce your case with vanilla SPM integration and then look for the `resource_accessor.swift` (I believe that's the name of the file they generate)

**Comment by danibachar** [comment]
A few things about the naming:

First, it does not solves the issue of un-used code.
Second, we would need to allow an API that support configuration for naming - such that you'd be able to give a custom name per target in the project i.e a map between target name and prefix.

I'm ok with that decision if that what you guys want, and just to clarify my understanding.

Instead of `disableBundleAccessors` we introduce `BundleAccessorOptions` defined as follow:
```swift
struct BundleAccessorOptions {
 typealias TargetName: String
 let enabled: Bool // indicates if to generate the accessors or not
 let accessorPropertyNamePrefixes: [TargetName: String] // mapping of prefixes for the class names
}
````

Use the prefixes (use the target name as default for missing targets) for generating the class names. 

*Please note, if we keep generating the classes and add a deprecation warning, and move to the property suggestion - that will not solve our issue for the upcoming future and change behaviour for users.

**Comment by fortmarek** [comment]
Next steps that we should do imho:
- Create a reproducible sample
- Reproduce the same setup with vanilla SPM integration
- Inspect what Apple is doing and follow their lead

I _think_ that they don't generate the objc resource accessor if the module doesn't have objc sources. Which is probably something we should do, too. That would also eliminate the generation of unused code. We have access to sources, so should be possible.

**Comment by fortmarek** [comment]
> I also understand now, that the objc accessors, are used by many (at least fixtures) to access a target resources externally - which I guess makes it useful in some cases...

We can also consider having slightly different resource accessor for SPM packages and for internal modules. The accessor for SPM packages should be inline with what Apple is doing whereas ours can diverge from that a bit (to allow the functionality of accessing resources from different modules)

**Comment by danibachar** [comment]
Hi again @pepicrft and @fortmarek - thanks a lot for your guidance and patience around this feature.
Regarding this [PR](https://github.com/tuist/tuist/pull/6146). I have comment on it, and migrated the relevant code to this one.
It was much simpler - I hope you are ok with that - here is a [link](https://github.com/tuist/tuist/pull/6146#issuecomment-2037593549) to the explanation I gave for the changes. 
TL;DR - explicit distinguish between internal and public accessors.

Let me know what do you think - and Ill close the other PR

File: Sources/TuistGenerator/Mappers/ResourcesProjectMapper.swift [hunk 12]
```diff
@@ -78,10 +85,9 @@ public class ResourcesProjectMapper: ProjectMapping { // swiftlint:disable:this
             sideEffects.append(sideEffect)
         }
 
-        if project.isExternal,
```
Comment (danibachar) [inline Sources/TuistGenerator/Mappers/ResourcesProjectMapper.swift:81]: Please note I removed the external flag, I think we should comply to SPM and generate those if any objc source files are in + resources...

**Comment by fortmarek** [comment]
Let me continue the conversation from the other [PR](https://github.com/tuist/tuist/pull/6146) here:

> It will introduce breaking changes though - wouldn't it? I mean we don't know if users are using the TargetNameResource.swift class for external swift packages.

We have never marketed that SPM resources can be available externally and I don't think our users should expect us to support that option when it's not available via vanilla integration. While technically a breaking change, I think it's something we can live with.

> I do think that changes https://github.com/tuist/tuist/pull/6124 are quite simple and comprehensive, keeping backwards compatibility, and are not that big?

I'd still redo the PR and remove the `ProjectDescription` changes. If we don't have a use-case for the added granularity, I'd skip it. We can still consider it but on its own and as a follow-up.

**Comment by danibachar** [comment]
Re-opened this [PR](https://github.com/tuist/tuist/pull/6146) to address this issue.
Closing his one in favor of the other
